_posts/2021-08-03-scalable-ets-counters.md

```diff
@@ -0,0 +1,222 @@
+benchmark results above show that both `set` and `ordered_set` tables
+get a significant scalability boost when the `decentralized_counter`
+option is activated. The `ordered_set` type receives a more
+substantial scalability improvement than the `set` type. Tables of the
+set type have a fixed number of locks for the hash table buckets. The
+`ordered_set` table type is implemented with a [contention adapting
+search tree](https://doi.org/10.1016/j.jpdc.2017.11.007) that
+dynamically changes the locking granularity based on how much
+contention is detected. This implementation difference explains the
+difference in scalability between `set` and `ordered_set`. The
+interested reader can find details about the `ordered_set`
+implementation in an [earlier blog
+post](https://blog.erlang.org/the-new-scalable-ets-ordered_set/).
+
+Worth noting is also that the Erlang VM that ran the benchmarks has
+been compiled with the configure option "`./configure
+--with-ets-write-concurrency-locks=256`". The configure option
+`--with-ets-write-concurrency-locks=256` changes the number of locks
+for hash-based ETS tables from the current default of 64 to 256 (256
+is currently the max value one can set this configuration option
+to). Changing the implementation of the hash-based tables so that one
+can set the number of locks per table instance or so that the lock
+granularity is adjusted automatically seems like an excellent future
+improvement, but this is not what this blog post is about.
+
+A centralized counter consists of a single memory word that is
+incremented and decremented with atomic instructions. The problem with
+a centralized counter is, of course, that modifications of the counter
+by multiple cores are serialized. This problem is amplified because
+frequent modifications of a single memory word by multiple cores cause
+a lot of expensive traffic in the [cache
+coherence](https://en.wikipedia.org/wiki/Cache_coherence)
+system. However, reading from a centralized counter is quite efficient
+as the reader only has to read a single memory word.
+
+When designing the decentralized counters for ETS, we have tried to
+optimize for update performance and scalability as most applications
+need to get the size of an ETS table relatively rarely. However, since
+there may be applications out in the wild that frequently call
+[`ets:info(Table, size)`](https://erlang.org/doc/man/ets.html#info-2)
+and [`ets:info(Table,
+memory)`](https://erlang.org/doc/man/ets.html#info-2), we have chosen
+to make decentralized counters optional. Another thing that might be
+worth keeping in mind is that the hash-based tables use a different
+heuristic to decide when the number of buckets shall grow and shrink
+when decentralized counters are activated. The hash table heuristics
+for growing and shrinking used together with decentralized counters
+tend to lead to tables that use slightly more memory than the default
+heuristics.
```

Comment:
Why was a different heuristics chosen? Is it faster, or necessary for the implementation of decentralized counters to work well?

Reply:
Perhaps it is wrong to say that the heuristics changed.

We use the same heuristics but instead of being based on the exact number of items in the table it is based on an estimate (i.e., the number of items protected by the current bucket lock times the number of buckets). This tend to lead to larger tables (at least for small tables) as growing is done more eagerly than shrinking. I will try to reformulate to make this more clear.

Reply:
Using the decentralized counters to get an estimate would be too slow as it would require going through all slots in the counter array.

Reply:
Fixed in b8fb1b3. Please have a look to see if you think it is better now.